### ISSUE TITLE
Classpath lookup discriminates on file extension (does not work with non .jar files)

### ISSUE THREAD

**Issue by jodersky** [lampepfl/dotty#13059]
## Compiler version

3.0.0 (but also tested with master from 2021-07-12)

## Problem

Embedding the compiler in a file not ending with `.jar` does not respect the user's classpath (when called with  `-usejavacp`)

The gist of it is that I have a fat jar bundled as an executable with a shell script header. Essentially, this file is both a valid shell script (starting with `#!/bin/sh`) and also a zip file, containing all of my application's classfiles. This is a very pleasant way to distribute final user programs which feel "static": single binary, no installation step required, no dependencies apart from the JVM.

Now, I'm trying to call the dotty compiler from this application, setting the "-usejavacp" flag, since I already have the whole scala lib bundled in my executable. Essentially, the executable is structured as follows:

```
#!/bin/sh
exec java  $JAVA_OPTS -cp "$0" '<main class>' "$@"
... contents of the fat jar
```

and the main method does the following:

```scala
def main(args: Seq[String]) = {
  val res = dotty.tools.dotc.Main.process(
    Array("-usejavacp") ++ args.toArray
  )
  if (res.hasErrors) sys.exit(1) else sys.exit(0)
}
```

The problem is that the compiler will give the following error:

> Could not find package scala from compiler core libraries.
> Make sure the compiler core libraries are on the classpath.

**unless my executable file has an extension of ".jar"** (which is not a very common thing on unix systems, as executables typically have no extension).

## Expectation

File extensions should not influence how the classpath is handled.

**Comment by jodersky** [comment]
Addendum: the setting `-Ylog-classpath` indeed confirms that the user classpath is empty if the executable (`$0`) in the script does not end with `.jar`, but is correctly detected if it does end with `.jar`.

**Comment by griggt** [comment]
The classpath is assembled in `PathResolver#Calculated.basis`

https://github.com/lampepfl/dotty/blob/cf6fa975413058c1efdcad2cc4a46513ff8f8aa2/compiler/src/dotty/tools/dotc/config/PathResolver.scala#L212-L221

and the `classesInExpandedPath(javaUserClassPath)` call leads to 

https://github.com/lampepfl/dotty/blob/cf6fa975413058c1efdcad2cc4a46513ff8f8aa2/compiler/src/dotty/tools/dotc/classpath/ClassPathFactory.scala#L53-L62

which results in two separate checks of the file extension (shown below).

The first happens in `AbstractFile.getDirectory`:

https://github.com/lampepfl/dotty/blob/cf6fa975413058c1efdcad2cc4a46513ff8f8aa2/compiler/src/dotty/tools/io/AbstractFile.scala#L36-L44

and the second in `ClassPathFactory.newClassPath`:

https://github.com/lampepfl/dotty/blob/cf6fa975413058c1efdcad2cc4a46513ff8f8aa2/compiler/src/dotty/tools/dotc/classpath/ClassPathFactory.scala#L74-L82

where the `isJarOrZip` extension method on AbstractFile is defined as:

https://github.com/lampepfl/dotty/blob/cf6fa975413058c1efdcad2cc4a46513ff8f8aa2/compiler/src/dotty/tools/dotc/classpath/FileUtils.scala#L23-L24

**Comment by som-snytt** [comment]
Reminiscent of https://github.com/scala/scala/pull/8102/files where jar is in exe file.

I don't remember offhand nsc.io vs reflect.io etc but this may fall under missed opportunity for code sharing.

**Comment by griggt** [comment]
Yes, looks like we need to pick up 8102 from scala/scala.  But how to adapt `AbstractFile.getDirectory` (without actually probing the file contents)?

**Comment by anatoliykmetyuk** [comment]
> Embedding the compiler in a file

It would be good to minimize away this step, or provide some clear instructions on how to do it.
